backend/routes/graph.py

```diff
@@ -29,166 +35,308 @@ def subgraph():
+@router.get("/indented-concept-list")
+async def indented_concept_list(request: Request, id: List[int] = Query(...)):   # id is a list of concept ids
+    return await indented_concept_list_post(request=request, id=id)
+
+
+@router.post("/indented-concept-list")
+async def indented_concept_list_post(request: Request, id: Union[List[int], None] = None) -> List:
     rpt = Api_logger()
     await rpt.start_rpt(request, params={'concept_ids': id})
 
     try:
-        sg = connected_subgraph_from_nodes(id, REL_GRAPH, REL_GRAPH_UNDIRECTED)
-        edges = [(str(e[0]), str(e[1])) for e in sg.edges]
-        await rpt.finish(rows=len(edges))
+        # new way to do this
+        paths = all_paths(REL_GRAPH, id)
+        tree = paths_as_indented_tree(paths)
+        await rpt.finish(rows=len(tree))
     except Exception as e:
         await rpt.log_error(e)
         raise e
-    return edges
+    return tree
 
 
-@router.get("/subgraph")
-async def subgraph(request: Request, id: List[int] = Query(...)):   # id is a list of concept ids
-    return await subgraph_post(request=request, id=id)
-
-
-@router.get("/hierarchy")
-def hierarchy(id: List[int] = Query(...)):   # id is a list of concept ids
+def all_paths(g: networkx.DiGraph, nodes: List[int]) -> List[List[int]]:
     """
-        couldn't figure out how to send the hierarchy in order of nodes with most
-        descendants first -- was able to sort it, but fastapi just resorted it in
-        key order before sending. going to try to do all the hierarchy stuff at the
-        front end instead. including the gap filling.
+    Creates a subgraph from g using nodes.
+    Identifies root and leaf nodes in this subgraph.
+    Generates all simple paths from each root to each leaf in the original graph g.
+    Returns the list of all such paths.
     """
-    sg = connected_subgraph_from_nodes(id, REL_GRAPH, REL_GRAPH_UNDIRECTED)
-    j = graph_to_json(sg)
-    return j
-    # return [json.dumps(j)]
-    # return Response(content=json.dumps(j), media_type='application/json')
-    # return JSONResponse(OrderedDict(j))
-    # return OrderedDict(j)
-
-
-
-def connected_subgraph_from_nodes(nodes, G, G_undirected=None, verbose=False):
-    # what this fails to do is fill in gaps if the subgraph is already connected
-    timer = get_timer('  connected_subgraph_from_nodes')
-    if not G_undirected:
-        verbose and timer(f'convert G to undirected')
-        G_undirected = G.to_undirected()
-
-    verbose and timer(f'subgraph for {len(nodes)} nodes')
+    sg = g.subgraph(nodes)
+    roots = [node for node, degree in sg.in_degree() if degree == 0]
+    leaves = [node for node, degree in sg.out_degree() if degree == 0]
+    paths = []
+    for root in roots:
+        for leaf in leaves:
+            _paths = list(nx.all_simple_paths(g, root, leaf))
+            for path in _paths:
+                if len(path) > 1:
+                    paths.append(path)
+    return paths
+
+
+def fill_in_gaps(G, nodes):
```

Comment:
`fill_in_gaps()`: I see how this works! I know that the full set of nodes is needed but I'm not sure how the frontend uses that full set, but that's OK for now.

Reply:
This is not actually being used yet. It's basically the same thing as `all_paths()`, but returns a subgraph instead of a path list. When we get to visualizing these hierarchies, we won't need to repeat concepts for every parent, so we'll use this.